scripts/utilities/param_space_exploration_utilities.py

```diff
@@ -1,85 +1,246 @@
-       without any communication, race conditions, or other problems
+# The "explore_parameter_space" function creates an array of subdirectories, each containing:
+#   - a run script (bash), generally called run.run
+#   - a parameter file (txt), generally caled run.par
+# Note that the multi-directory structure is useful for several reasons:
+#   1. The output for each run stays in its own subdirectory, attached to the appropriate input files
+#   2. If multiple processors are available, they may simultaneously run separate instances of rescal-snow, one per directory,
+#       without any communication, race conditions, or other problems
 
-"""
+# Set up a bunch of runs to go in parallel
 
-import rescal_utilities
 import numpy as np
 import datetime
 import os
+import itertools
+import subprocess
+import glob
+import rescal_utilities
+
 
 def explore_parameter_space(output_root, executable_location, parameter_ranges, fixed_parameters, n_runs):
-	# Take in a dictionary of parameter ranges to explore,
-	#   e.g. { 'Lambda_S' : [0.001, 0.2],
-	#          'Lambda_F' : [3, 10] }
-
-	# Setup n_runs runs to explore those, under directory output_ro
-	# Rescal and genesis executables are located in directory executable_location
-
-	if ( os.path.isdir(output_root) == False):
-		os.mkdir(output_root)
-
-	# Get some lists of parameter combos to explore
-	parameter_combos = _uniform_random_search(parameter_ranges, n_runs)
- 	
-	# Write parameter files for each set of parameters
-	now = datetime.datetime.now()
-	for run in range(n_runs):
-		run_output_dir 		= output_root + "/" + \
-				str(now.year) + str(now.month) + str(now.day) + "_" + str(run)
-
-		# Avoid overwriting previous results or using a messy directory
-		while os.path.isdir(run_output_dir):
-			run_output_dir += "A"
-
-		this_run = rescal_utilities.Design_a_run()
-		this_run.set_parameters(parameter_combos[run])
-		this_run.set_parameters({'rescallocation' : executable_location,
-			'parfile': 'run.par'})
-		this_run.set_parameters(fixed_parameters)
-		this_run.set_directory(run_output_dir)
-
-		os.mkdir(run_output_dir)
-		this_run.write()
+    # Take in a dictionary of parameter ranges to explore,
+    #   e.g. { 'Lambda_S' : [0.001, 0.2],
+    #          'Lambda_F' : [3, 10] }
+
+    # Setup n_runs runs to explore those, under directory output_ro
+    # Rescal and genesis executables are located in directory executable_location
+
+    if ( os.path.isdir(output_root) == False):
+        os.mkdir(output_root)
+
+    # Get some lists of parameter combos to explore
+    parameter_combos = _uniform_random_search(parameter_ranges, n_runs)
+
+    # Write parameter files for each set of parameters
+    now = datetime.datetime.now()
+    for run in range(n_runs):
+        run_output_dir          = output_root + "/" + \
+                        str(now.year) + str(now.month) + str(now.day) + "_" + str(run)
```

Comment:
string.format

Reply:
Fixed, but this is Kelly's code.

Reply:
Thanks. I think it's affected by the PR, so I mentioned it. My apologies if that's wrong.